### DISCUSSION TITLE
How to Include Custom Folder Structure in MSI Installation in appdata/temp folder

### DISCUSSION THREAD

**Discussion by surajk46** [oleg-shilo/wixsharp#1674] [Q&A] [answered]
Hi Oleg
I need to place additional files (DLLs and other assemblies as it may have some dependency like xml,config files,some msi files ) into a specific folder structure. These files are helper files to perform other actions like validations and other stuff. These files will not be installed but used as temp files for other activity. As I understand correctly, all such files are copied under temp folder created for installer. My observation is that If I use Project.DefaultRefAssembly.Add() then this assemblies are copied under this temp folder. However, I would like to customize this behavior. Instead of placing the files directly in the temporary folder, I want to create a specific folder structure and place these assemblies with their dependency in the correct subfolders

I wanted a folder structure like below
![image (1)](https://github.com/user-attachments/assets/60d17bea-67cc-4070-9e9c-174e0ba75b67)

**Comment by oleg-shilo** [comment, marked as answer]
Unfortunately, you cannot do that directly.
The added `DefaultRefAssembly` files are extracted by MSI/WiX in the destination temp folder which we have no control over. It's not possible to change.

However, if you only want to have some files present and used for a short time (e.g. deployment validation) you can always embed these files in msi but without installing them.

Then you can copy them to the destination of your choice during the installation from the special custom action (e.g. `project.Load` event). 
And you can simply delete these files when you do not heed them any more.

Embedding the files is easy. Either as a resource for your assembly that implements `project.Load` or as a true MSI entity `new Binary("path to your file")`. Then you can extract with the `public static byte[] GetEmbeddedData(this Session session, string binary)`.

**Reply by surajk46** [reply]
Hi, thanks for the response! This solution works quite well for me.

I added the ZIP file using:

`project.AddBinary(new Binary("../HelperFolder/HelperFiles.zip"));`

Then, in the UIInitialize event, I extract the file with the following code and unzip it:
            
```
byte[] fileData = e.Session.GetEmbeddedData("HelperFiles.zip");
string basePath = AppDomain.CurrentDomain.BaseDirectory; //returns apdata/temp folder 
string filePath = basePath + "\\Helper.zip";
System.IO.File.WriteAllBytes(filePath, fileData); // Extract ZIP
```

However, I'm noticing that it takes quite a long time to convert the byte[] data into a file. 
when i double click on msi firsly it take around 1 min to show the first UI, But after this the folder structure from zip is extracted.

**Reply by oleg-shilo** [reply]
What you did is quite a simple modification. You just added a file in the MSI table. That's it. You did not schedule any MSI actions or establish some relationships between MSI entities. Since you opted to do all the activities on the zip file to be done not by MSI but by your custom action you have practically done nothing to your MSI. 

This means that the delay is caused by the MSI unrelated circumstances. It can be anything: too large zip file. your AV trying to scan your zip file, etc.

While you have done everything right way, there's not much you can do to address this delay. Unfortunatelly.

**Reply by surajk46** [reply]
Hi Oleg,
Your solution works quite well for me, but I have encountered an issue. I need the same folder structure or helper files (including some third-party DLLs) for the uninstallation process. However, during uninstallation, I no longer have access to the temporary folder.
